ICU-724 Deal with scroll pop from single image previews

jwilander opened the pull request:
#### Summary
There were a few separate issues:
* Viewport width was starting at 0, so our initial height would be 0 even if we had the file height already (fixed by first commit)
* The loading indicator div was not the same height as the final image div, it was off by 10-15 pixels (fixed by first commit)
* In cases where the app was waiting on the file info to load, there was no scrolling of the post list to deal with the scroll pop from these mounting (fixed by second commit)

#### Ticket Link
https://mattermost.atlassian.net/browse/ICU-724

jwilander commented on components/single_image_view.jsx:
```diff
@@ -102,36 +106,29 @@ export default class SingleImageView extends React.PureComponent {
         }
 
-        return {previewWidth, previewHeight};
+        return previewHeight;
```
I don't really think computing and setting the width was useful at all, it looks like the CSS was handling those cases fine based on my testing

saturninoabril replied:
It's useful during initial load of image where image should scale up/down based on available browser window. Actually such feature was lost after we dealt with the edge case like (1) very tall but very thin, or (2) very wide but very short image.

Take an example of full AboutModal screenshot, on small window it didn't resize. Instead, it scaled down only based on max preview height then cropped the right part of the image.
![screen shot 2018-02-08 at 9 30 05 pm](https://user-images.githubusercontent.com/5334504/35975331-53808f8e-0d17-11e8-94af-9f7632a823c0.png)


@asaadmahmood has a separate fix for that here https://github.com/mattermost/mattermost-webapp/pull/743

jwilander replied:
Ok cool, if #743 fixes that then let's take that as well

jwilander commented on components/single_image_view.jsx:
```diff
@@ -190,13 +186,15 @@ export default class SingleImageView extends React.PureComponent {
-                        <div className={`image-loaded ${fadeInClass}`}>
+                        <div
+                            className={`image-loaded ${fadeInClass}`}
+                            style={imageContainerStyle}
```
Move the height setting down one level and set it always to ensure that height will not change

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0b355fe992a284fd7.spinmint.com

Test Admin Account: Email: `user-0@sample.mattermost.com` | Password: `user-0`

Test User Account: Email: `user-1@sample.mattermost.com` | Password: `user-1`

Instance ID: i-0b355fe992a284fd7

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0df37853b2f39336c.spinmint.com

Test Admin Account: Email: `user-0@sample.mattermost.com` | Password: `user-0`

Test User Account: Email: `user-1@sample.mattermost.com` | Password: `user-1`

Instance ID: i-0df37853b2f39336c

saturninoabril commented on components/single_image_view.jsx:
```diff
@@ -47,6 +48,9 @@ export default class SingleImageView extends React.PureComponent {
         this.loadImage(this.props.fileInfo);
+
+        // Timeout used to delay execution until after current render cycle
+        setTimeout(postListScrollChange, 0);
```
I never thought about this one, cool!

saturninoabril commented on components/single_image_view.jsx:
```diff
@@ -168,8 +162,7 @@ export default class SingleImageView extends React.PureComponent {
 
             fadeInClass = 'image-fade-in';
-            imageLoadedDimension = {cursor: 'pointer'};
-            imageContainerDimension = {};
```
I intentionally clear this style after the image is loaded so that CSS will handle image responsiveness on resize of browser or like expanding/shrinking of RHS.

jwilander replied:
If we want that we need to be 100% sure the CSS height matches with the height we set previously, which it wasn't

saturninoabril replied:
I did some check and indeed you're correct.

mattermod commented:
Spinmint test server destroyed

jwilander commented:
Alright I rebased it to work with the changes from #743 but after more testing I realized my changes did not fix scroll pop when there was a new messages indicator present and files that had not been loaded.

This is a **very** difficult problem to fix perfectly without some major overhauls on how this is all architected. That said, I do believe I've implemented a manageable fix that only has a minor visual downside and should work well enough in most cases.

What I did was add a div before the file info gets loaded that has a minimum height that is the maximum height potential height of the image. When the file info does get loaded and we know the actual image height, then it shrinks to the right height (or stays the same if it's too tall). The aforementioned downside is that any time an image that is below the maximum height is loaded **for the first time since page load** (subsequent loads will work perfectly) you'll briefly see a 385 pixel blank space before the file info loads.

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-038d36bf2368f9001.spinmint.com

Test Admin Account: Email: `user-0@sample.mattermost.com` | Password: `user-0`

Test User Account: Email: `user-1@sample.mattermost.com` | Password: `user-1`

Instance ID: i-038d36bf2368f9001

esethna commented:
Thanks @jwilander, makes sense. I tested this out and it's definitely a major improvement. On slow connections the div isn't great for small images, but after that first load it looks nice. I think we take this for 4.7 and then ICU has scrolling overhaul planned for an upcoming release.

mattermod commented:
Spinmint test server destroyed
